sycl/unittests/helpers/include/helpers/PiMock.hpp

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This mini-library provides facilities to test the DPC++ Runtime behavior upon
+// specific results of the underlying low-level API calls. By exploiting the
+// Plugin Interface API, the stored addresses of the actual plugin-specific
+// implementations can be overwritten to point at user-defined mock functions.
+//
+// While this could be done manually for each unit-testing scenario, the library
+// aims to rule out the boilerplate, providing helper APIs which can be re-used
+// by all such unit tests. The test code stemming from this can be more consise,
+// with little difference from non-mock classes' usage.
+//
+// The following unit testing scenarios are thereby simplified:
+// 1) testing the DPC++ RT management of specific PI return codes;
+// 2) coverage of corner-cases related to specific data outputs
+//    from underlying runtimes;
+// 3) testing the order of PI API calls;
+// ..., etc.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/pi.hpp>
+#include <CL/sycl/device_selector.hpp>
+#include <CL/sycl/platform.hpp>
+#include <CL/sycl/queue.hpp>
+#include <detail/platform_impl.hpp>
+
+#include <functional>
+#include <unordered_map>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace unittest {
+
+namespace detail = cl::sycl::detail;
+namespace RT = detail::pi;
+
+template <typename T> struct DispatchHelper;
+
+template <typename R, typename... Args> struct DispatchHelper<R(Args...)> {
+  using type = std::function<R(Args...)>;
+};
+
+template <detail::PiApiKind Kind> struct PiFuncType;
+
+#define _PI_API(api)                                                           \
+  template <> struct PiFuncType<detail::PiApiKind::api> {                      \
+    using type = typename DispatchHelper<decltype(api)>::type;                 \
+  };
+#include <CL/sycl/detail/pi.def>
+
+#undef _PI_API
+
+struct PiDispatch {
+#define _PI_API(api)                                                           \
+  typename DispatchHelper<decltype(api)>::type mock_##api = [](auto...) {      \
+    std::cerr << "Unexpected PI call: " << #api << "\n";                       \
+    std::terminate();                                                          \
+    return PI_SUCCESS;                                                         \
+  };
+
+#include <CL/sycl/detail/pi.def>
+
+#undef _PI_API
+};
+
+extern std::unordered_map<backend, PiDispatch> *GDispatchTables;
+
+void hijackPlugins();
+
+/// Overwites PI call for a single backend.
+///
+/// \param F is any callable (function or lambda), that will be used instead of
+/// default PI call handler.
+template <detail::PiApiKind PiKind, backend Backend>
+void redefineOne(const typename PiFuncType<PiKind>::type &F) {
```

Comment:
It seems we often need to add a very simple logic like counting number of calls or checking if one of the arguments is correct. While this is not too difficult to redefine some of the APIs, there are set of APIs which would would require copy-pasting of default redefine body + a couple of lines to specific kernel needs.

So, the suggestion is to add something like append or prepend functionality, so a user(test writer) can provide a function which runs in addition(before or after) the function[s] which are already set for an API.